### PR TITLE
MsSqlHook: implement _generate_insert_sql

### PR DESCRIPTION
**Pull request by jplauri**
Related: https://github.com/apache/airflow/discussions/19583

For prepared SQL statements, MSSQL uses `?` rather than `%s` for parameter markers. Currently, [MsSqlHook](https://airflow.apache.org/docs/apache-airflow-providers-microsoft-mssql/stable/_api/airflow/providers/microsoft/mssql/hooks/mssql/index.html) does not provide an implementation of `_generate_insert_sql` used by `insert_rows` making it impossible to use for instance [GenericTransfer](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/generic_transfer/index.html) with the said hooks.

As written, the PR implements `_generate_insert_sql` for MsSqlHook but in a limited way. That is, it will only allow us to do INSERTs (so when `replace=False`) but not UPSERTs (when `replace=True`). I would love to get some feedback and ideas for implementing the UPSERT even though it is not useful for my current use-case.

Note that MSSQL does not support anything like "REPLACE INTO" for UPSERT. Instead, there seem to be two alternatives:

- Try to do an update first and check `@@ROWCOUNT`: in case nothing was updated, do an insert. This is explained [here by Aaron Bertrand](https://sqlperformance.com/2020/09/locking/upsert-anti-pattern).

- Use `MERGE`, many hits on StackOverflow such [this one](https://stackoverflow.com/q/2479488/551375) explaining the syntax.

It seems that Aaron B. argues for the pattern in the first link because of several issues with `MERGE` (see [list of issues](https://www.mssqltips.com/sqlservertip/3074/use-caution-with-sql-servers-merge-statement/)).

I spent a bit of time trying to implement either approach here, but got stuck after realizing it's not that straightforward to re-use parameter markers `?` from `values` from pyodbc `cursor.execute(sql, values)`. In any case, I was attempting to draw some inspiration from `_generate_insert_sql` of [PostgresHook](https://airflow.apache.org/docs/apache-airflow-providers-postgres/stable/_modules/airflow/providers/postgres/hooks/postgres.html#PostgresHook). Thus, this might need some work beyond just implementing `_generate_insert_sql` unless I'm missing something.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by dstandish** [comment]
@jplauri regarding your question about merge vs insert, i think merge is the right call.  That list of issues  is pretty ancient now.  Merge was implemented in 2008 and by now, I think most of the meaningful kinks have been ironed out.  I would bet that the rowcount approach would be slow since it limits you to row-by-row.

But I don't think you need to tackle that in this PR.  One thing at a time.  Can do this limited case first.

**Comment by jplauri** [comment]
> @jplauri regarding your question about merge vs insert, i think merge is the right call. That list of issues is pretty ancient now. Merge was implemented in 2008 and by now, I think most of the meaningful kinks have been ironed out. I would bet that the rowcount approach would be slow since it limits you to row-by-row (where as with merge you could, e.g. load into a temp table and merge into target with deduping logic too)
> 
> But I don't think you need to tackle that in this PR. One thing at a time. Can do this limited case first.

Thanks! And yes, I would be fine with this limited solution too. I mean, GenericTransfer is meant to be used for small datasets anyway as it works in-mem, so I can imagine there are cases where doing say truncate & insert all is just fine instead of replace.

**Comment by dstandish** [comment]
>  I mean, GenericTransfer is meant to be used for small datasets anyway as it works in-mem

Yeah I wasn't aware of that, this is my first look at generic transfer. 

> so I can imagine there are cases where doing say truncate & insert all is just fine instead of replace

Yeah this is one of the challenges with trying to build an interface like this.  Replace can mean "replace rows (i.e. when collision by PK)" or "replace table".  And in the base implementation, it means by row.  But you are talking by table.  So there's ambiguity.

**Comment by dstandish** [comment]
Oh... _generate_insert_sql already _is_ row-by-row.  I assumed it is row-based. Yeah in that case def small datasets / limited use cases.

**Comment by jplauri** [comment]
@dstandish Would you be able to advise me further here, i.e., is there something more the PR should have? Should we just wait for a review/merge, or what's the process? I'm in no rush, but I believe this addition would be helpful for some people.

**Comment by dstandish** [comment]
> @dstandish Would you be able to advise me further here

i'll review today

**Review by dstandish — CHANGES_REQUESTED** [review]
Please add a test for this

**Comment by jplauri** [comment]
@dstandish Not really sure if I got it right - I basically copied from TestPostgres with only superficial understanding. But intuitively it makes sense - running `test_mssql_to_mssql` will fail unless `_generate_insert_sql` is at least syntactically correct. What do you think?

**Comment by uranusjr** [comment]
Let’s see if tests can run now.

File: tests/operators/test_generic_transfer.py [hunk 1]
```diff
@@ -26,6 +26,7 @@
 from airflow.models.dag import DAG
 from airflow.operators.generic_transfer import GenericTransfer
 from airflow.providers.mysql.hooks.mysql import MySqlHook
+from airflow.providers.microsoft.mssql.hooks.mssql import MsSqlHook
```
Comment (eladkal) [inline tests/operators/test_generic_transfer.py]:
```suggestion
from airflow.providers.microsoft.mssql.hooks.mssql import MsSqlHook
from airflow.providers.mysql.hooks.mysql import MySqlHook
```

To fix static tests

**Review by eladkal — DISMISSED** [review]

**Comment by uranusjr** [comment]
I merged main for #21221.

**Review by uranusjr — APPROVED** [review]

**Comment by potiuk** [comment]
tests /static checks failing

**Comment by jplauri** [comment]
> Please add a test for this

Should this be good to go now?

**Comment by potiuk** [comment]
I rebased it - let's see. @dstandish ?

**Comment by jplauri** [comment]
Just to avoid this going stale as I think it could be helpful. @dstandish do you think some work is still required?

**Comment by jplauri** [comment]
> Please add a test for this

Would you have a chance to revisit this PR? Thanks!

**Comment by eladkal** [comment]
mm in Sqlite we removed  `_generate_insert_sql` in the hook see https://github.com/apache/airflow/pull/25939
could same approch be relvant here?
cc @alexandermalyga

**Comment by dstandish** [comment]
very sorry this slipped off the radar @jplauri.  please do check out @eladkal's commant and i'll watch for the update

**Pull request closed without merging**